lib/client-request.js

```diff
@@ -0,0 +1,176 @@
+    this._responseHeadBuffer = null
+    this._responseBodyBuffer = null
+
+    this._pendingFinal = null
+
+    if (onresponse) this.once('response', onresponse)
+
+    socket
+      .on('close', this._onclose.bind(this))
+      .on('data', this._ondata.bind(this))
+      .on('end', this._onend.bind(this))
+      .on('error', this._onerror.bind(this))
+  }
+
+  _header () {
+    let h = `${this.method} ${this.path} HTTP/1.1\r\n`
+
+    for (const name of Object.keys(this.headers)) {
+      const n = name.toLowerCase()
+      const v = this.headers[name]
+
+      h += `${httpCase(n)}: ${v}\r\n`
+    }
+
+    h += '\r\n'
+
+    return h
+  }
+
+  _final (cb) {
+    if (this.headersSent === false) this.flushHeaders()
+
+    this._pendingFinal = cb
+  }
+
+  _ondata (data) {
+    if (!this._responseHeadComplete) {
+      const blankLine = Buffer.from([13, 10, 13, 10])
+      const blankLineIndex = data.indexOf(blankLine)
+      const hasBlankLine = blankLineIndex !== -1
+
+      const h = hasBlankLine ? data.subarray(0, blankLineIndex) : data
+
+      if (this._responseHeadBuffer !== null) {
+        this._responseHeadBuffer = Buffer.concat([this._responseHeadBuffer, h])
+      } else {
+        this._responseHeadBuffer = h
+      }
+
+      if (hasBlankLine) {
+        this._responseHeadComplete = true
+        this._readHead(this._responseHeadBuffer)
+        data = data.subarray(blankLineIndex + blankLine.byteLength)
+      }
+    }
+
+    if (this._responseHeadComplete) this._readBody(data)
+  }
+
+  _readHead (head) {
+    const h = head.toString().split('\r\n')
+
+    let [, statusCode, statusMessage] = h[0].split(' ')
+    if (!statusCode || !statusMessage) return this._socket.destroy()
+
+    statusCode = +statusCode
+
+    const headers = {}
+
+    for (let i = 1; i < h.length; i++) {
+      const [name, value] = h[i].split(': ')
+      headers[name.toLowerCase()] = value
+    }
+
+    this._response = new HTTPIncomingMessage(this._socket, headers, { statusCode, statusMessage })
+
+    this.emit('response', this._response)
+  }
+
+  _readBody (data) {
+    const len = Number(this._response.getHeader('content-length')) || 0
+    const chunked = this._response.getHeader('transfer-encoding') === 'chunked'
+    const chunkedTail = Buffer.from([48, 13, 10, 13, 10]) // '0\r\n\r\n'
+
+    if (this._responseBodyBuffer !== null) {
+      this._responseBodyBuffer = Buffer.concat([this._responseBodyBuffer, data])
+    } else {
+      this._responseBodyBuffer = data
+    }
+
+    const hasChunkedTail = chunked && this._responseBodyBuffer.indexOf(chunkedTail) !== -1
+    if (hasChunkedTail) {
+      let message = Buffer.from([])
+
+      let readingMessage = false
+      let hits = 0
+
+      for (let i = 0; i < this._responseBodyBuffer.byteLength; i++) {
+        const b = this._responseBodyBuffer[i]
+
+        if (readingMessage && hits === 0 && b !== 13) {
+          message = Buffer.concat([message, Buffer.from([b])])
```

Comment:
This is indeed the bottleneck as it builds up the buffer in `O(n^2)` time. Instead of incrementally appending bytes to the message buffer, you can keep track of where the current chunk started and then grab the entire chunk using `subarray(start, i)` as soon as the chunk ends.

Reply:
This is essentially what https://github.com/holepunchto/bare-http1/blob/902e63cf02ad24341e5c5156a66e5bfa27bb5279/lib/server-connection.js#L53-L55 does as well, but it also splits the remainder so the current chunk always starts at 0. That's less state to keep track of.

Reply:
great, I am using `Buffer.concat` just once and it has improved performance a lot.